### PR TITLE
Finished brush tool.

### PR DESCRIPTION
**Pull request by abdominator**

### CODE REVIEW

File: src/Editor/Editor.cpp [hunk 1]
```diff
@@ -28,7 +28,6 @@
 #include <fstream>
 #include <Utility/Log.hpp>
 
-
 ImGuizmo::OPERATION currentOperation = ImGuizmo::TRANSLATE;
 Editor::Editor() {
```
Comment (NinjaDanz3r) [inline src/Editor/Editor.cpp]: MINOR Member variable 'Editor::currentEntity' is not initialized in the constructor. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)

File: src/Engine/Entity/Entity.cpp [hunk 2]
```diff
@@ -336,6 +336,9 @@ unsigned int Entity::GetUniqueIdentifier() const {
     return uniqueIdentifier;
 }
 
+void Entity::SetDirection(glm::vec3 rot) {
```
Comment (NinjaDanz3r) [inline src/Engine/Entity/Entity.cpp]: MINOR The function 'SetDirection' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 3 issues
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/Editor/Editor.cpp (hunk 1 above)
Comment (NinjaDanz3r) [inline src/Editor/Editor.cpp]: MINOR Member variable 'Editor::currentEntity' is not initialized in the constructor. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)

File: src/Engine/Entity/Entity.cpp [hunk 3]
```diff
@@ -336,6 +334,9 @@ unsigned int Entity::GetUniqueIdentifier() const {
...
```
Comment (NinjaDanz3r) [inline src/Engine/Entity/Entity.cpp]: MINOR The function 'SetDirection' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 3 issues
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/Editor/Editor.cpp (hunk 1 above)
Comment (NinjaDanz3r) [inline src/Editor/Editor.cpp]: MINOR Member variable 'Editor::currentEntity' is not initialized in the constructor. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AuninitMemberVar)

File: src/Engine/Entity/Entity.cpp (hunk 3 above)
Comment (NinjaDanz3r) [inline src/Engine/Entity/Entity.cpp]: MINOR The function 'SetDirection' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

File: src/Editor/Editor.cpp [hunk 4]
```diff
@@ -609,6 +560,192 @@ void Editor::Focus() {
...
+            bool intersect = false;
+            glm::vec3 last_p0;
+            glm::vec3 last_p1;
+            glm::vec3 last_p2;
+            glm::vec3 p0;
+            glm::vec3 p1;
+            glm::vec3 p2;
+
+            mousePicker.Update();
+            mousePicker.UpdateProjectionMatrix(cameraEntity->GetComponent < Component::Lens>()->GetProjection(glm::vec2(MainWindow::GetInstance()->GetSize().x, MainWindow::GetInstance()->GetSize().y)));
```
Comment (andens) [inline src/Editor/Editor.cpp]: Spacing around angle bracket.

File: src/Editor/Editor.cpp [hunk 5]
```diff
@@ -609,6 +560,192 @@ void Editor::Focus() {
...
+
+    ImGuiIO& io = ImGui::GetIO();
+    // Draw the actual widget.
+    ImGuizmo::SetRect(currentEntityMatrix[0][0], 0, io.DisplaySize.x, io.DisplaySize.y);
+    glm::mat4 deltaMatrix = glm::mat4();
+    ImGuizmo::Manipulate(&viewMatrix[0][0], &projectionMatrix[0][0], currentOperation, imguizmoMode, &currentEntityMatrix[0][0], &deltaMatrix[0][0]);
+
+    if (ImGuizmo::IsUsing()) {
+        switch (currentOperation) {
+        case ImGuizmo::TRANSLATE: {
```
Comment (andens) [inline src/Editor/Editor.cpp]: Indent case blocks.

File: src/Editor/Editor.hpp [hunk 6]
```diff
@@ -18,111 +18,132 @@ struct GLFWcursor;
 
 /// Handles the main editor.
 class Editor {
-    public:
```
Comment (andens) [inline src/Editor/Editor.hpp]: Wrong indentation of the class contents.

File: src/Editor/GUI/Editors/EntityEditor.hpp [hunk 7]
```diff
@@ -6,6 +6,7 @@
 #include <functional>
 #include <Engine/Entity/Entity.hpp>
 #include <imgui.h>
+#include <Engine/Geometry/AssetFileHandler.hpp>
```
Comment (andens) [inline src/Editor/GUI/Editors/EntityEditor.hpp]: There is no other change in this file (the header) so we should prefer to place this include in the cpp file instead.

File: src/Engine/Entity/Entity.cpp [hunk 8]
```diff
@@ -336,6 +334,9 @@ unsigned int Entity::GetUniqueIdentifier() const {
...
+    rotation = rot;
```
Comment (andens) [inline src/Engine/Entity/Entity.cpp]: There has been other changes to `GetDirection` which returns the direction in world space. We should be consistant and set world space as well. As of now this method sets the direction in the local space of the entity.

File: src/Engine/Entity/Entity.hpp [hunk 9]
```diff
@@ -241,13 +241,24 @@ class Entity {
         /**
          * @param UID the entity's unique identifier to be set
          */
+
+        ENGINE_API void SetDirection(glm::vec3 rot);
```
Comment (andens) [inline src/Engine/Entity/Entity.hpp]: This has been placed between the doxygen comment of `SetUniqueIdentifier` and that method. This should be relocated and have its own documentation as well.

File: src/Engine/Util/RayIntersection.cpp [hunk 10]
```diff
@@ -61,3 +61,41 @@ bool RayIntersection::RayOBBIntersect(const glm::vec3& rayOrigin, const glm::vec
         outputDistance = tMax;
     return true;
 }
+
+bool RayIntersection::TriangleIntersect(glm::vec3 origin, glm::vec3 direction, glm::vec3 p0, glm::vec3 p1, glm::vec3 p2, float & distance) {
```
Comment (andens) [inline src/Engine/Util/RayIntersection.cpp]: Ampersand goes to the left.

File: src/Engine/Util/RayIntersection.hpp [hunk 11]
```diff
@@ -28,4 +28,21 @@ class RayIntersection {
             const Video::AxisAlignedBoundingBox& meshData,
             const glm::mat4& modelMatrix,
             float &outputDistance) const;
+       
+        /// Checks whether a ray intersects with a triangle.
+        /**
+        * @param rayOrigin The position the ray originates from.
```
Comment (andens) [inline src/Engine/Util/RayIntersection.hpp]: Misaligned asterisks.

File: src/Engine/Util/RayIntersection.hpp [hunk 12]
```diff
@@ -28,4 +28,21 @@ class RayIntersection {
...
+        * @param p2 The third vertex of the triangle being tested.
+        * @param distance The distance to the entity that is tested to be used as a return value check.
+        * @return Whether or not the ray intersects with the triangle that is tested against.
+        */
+        ENGINE_API bool TriangleIntersect(glm::vec3 rayOrigin,
+            glm::vec3 rayDirection,
+            glm::vec3 p0,
+            glm::vec3 p1,
+            glm::vec3 p2,
+            float &distance);
```
Comment (andens) [inline src/Engine/Util/RayIntersection.hpp]: Ampersand goes to the left.

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**